In [1]:
# This Python 3 environment comes with many helpful analytics libraries installed
# It is defined by the kaggle/python docker image: https://github.com/kaggle/docker-python
# For example, here's several helpful packages to load in 

import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)

# Input data files are available in the "../input/" directory.
# For example, running this (by clicking run or pressing Shift+Enter) will list all files under the input directory

import os
for dirname, _, filenames in os.walk('/kaggle/input'):
    for filename in filenames:
        print(os.path.join(dirname, filename))

# Any results you write to the current directory are saved as output.

/kaggle/input/restaurantreviews/Restaurant_Reviews.tsv


**Importing Modules** 

In [2]:
# Importing the libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import re
import nltk
from nltk.stem.porter import PorterStemmer

In [3]:
# Importing the dataset
df = pd.read_csv('../input/restaurantreviews/Restaurant_Reviews.tsv', delimiter = '\t', quoting = 3)

In [4]:
df.head()

Review  Liked
0                           Wow... Loved this place.      1
1                                 Crust is not good.      0
2          Not tasty and the texture was just nasty.      0
3  Stopped by during the late May bank holiday of...      1
4  The selection on the menu was great and so wer...      1

In [5]:
#df['Review'][0]

**We will be keeping only letters and Removing numbers and Puncuations **

In [6]:
#review=re.sub('[^a-zA-Z]',' ',df['Review'][0])

In [7]:
#review

We can see now that we have removed puncuation from the first review

**Converting all the letters in the review into small case**

In [8]:
#review=review.lower()

In [9]:
#review

So we can see that the all the letters of the review column are converted to small case 

**Removing non Significant words (Stop Words)**

In [10]:
nltk.download('stopwords')
from nltk.corpus import stopwords


[nltk_data] Downloading package stopwords to /usr/share/nltk_data...
[nltk_data]   Package stopwords is already up-to-date!


Converting the review into list of words 

In [11]:
#review=review.split()

In [12]:
#ps=PorterStemmer()

In [13]:
#review

We Can see now that review is list of four words

**Stemming to keep only the root word**

In [14]:
#review=[ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
#review

We can see that after stemming Loved was converted to Love in out list of words review

**Joining the words in the list **

In [15]:
#review=' '.join(review)

In [16]:
#review

We can see that all the words in the list are joined and we have used space between the words

**Now lets clean data for whole data set using a loop**

In [17]:
corpus=[]
for i in range(0,1000):
    review=re.sub('[^a-zA-Z]',' ',df['Review'][i])
    review=review.lower()
    review=review.split()
    ps=PorterStemmer()
    review=[ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review=' '.join(review)
    corpus.append(review)

In [18]:
corpus[0:5]

['wow love place',
 'crust good',
 'tasti textur nasti',
 'stop late may bank holiday rick steve recommend love',
 'select menu great price']

We have the new list corpus which has all the words from the 1000 reviews.Using a for loop all the process like removing puncuations,making all words lower case and removing stop words and joining done in one stage 

**Creating Bag of Words Model**

In [19]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer()
X=cv.fit_transform(corpus).toarray()

In [20]:
X.shape

(1000, 1565)

We can remove to non frequent words by using max_features.In this case there are 1565 words in the word vector matrix.We can use word max as 1500 which would effectively remove 65 words from the spare word vector matrix

In [21]:
from sklearn.feature_extraction.text import CountVectorizer
cv=CountVectorizer(max_features=1500)
X=cv.fit_transform(corpus).toarray()

In [22]:
X.shape

(1000, 1500)

In [23]:
y=df.iloc[:,1].values

In [24]:
y.shape

(1000,)

**Splitting data into train and test Data**

In [25]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,test_size=0.15)

**Scaling the data **

In [26]:
from sklearn.preprocessing import StandardScaler
sc_X=StandardScaler()
X_train=sc_X.fit_transform(X_train)
X_test=sc_X.fit_transform(X_test)

**Fitting the Naive Bayes model to the training data**

In [27]:
from sklearn.naive_bayes import GaussianNB
classifier=GaussianNB()
classifier.fit(X_train,y_train)

GaussianNB(priors=None, var_smoothing=1e-09)

**Predicting the test results **

In [28]:
y_pred=classifier.predict(X_test)

* **Making the confusion matrix,Classification report and Accuracy Score**

In [29]:
from sklearn.metrics import confusion_matrix
cm=confusion_matrix(y_test,y_pred)
cm

array([[62,  5],
       [38, 45]])

In [30]:
from sklearn.metrics import confusion_matrix,classification_report,accuracy_score
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       0.62      0.93      0.74        67
           1       0.90      0.54      0.68        83

    accuracy                           0.71       150
   macro avg       0.76      0.73      0.71       150
weighted avg       0.77      0.71      0.71       150



In [31]:
print(accuracy_score(y_test,y_pred))

0.7133333333333334
